src/main/java/hudson/remoting/Request.java

```diff
@@ -234,11 +234,14 @@ public boolean isDone() {
             public RSP get() throws InterruptedException, ExecutionException {
                 synchronized(Request.this) {
                     try {
-                        while(response==null) {
+                        while(response==null && !channel.isInClosed()) {
                             if (isCancelled()) {
                                 throw new CancellationException();
                             }
-                            Request.this.wait(); // wait until the response arrives
+                            Request.this.wait(30*1000); // wait until the response arrives
```

Comment:
Waits w/o timeout are evil, but such hardcoded timeout is not applicable in the case of network-robust instances